[export] Upstream support of (tensor, tensor list) in op returns.

zhxchen17 opened the pull request:
Summary:
Upstreaming from internal to oss.
Diff: D49710320

Test Plan: buck2 build mode/opt sigmoid/inference/test_gpu:package_gen

Differential Revision: D50577490




cc @avikchaudhuri @gmagogsfm @tugsbayasgalan @angelayi

SherlockNoMad commented:
test case plz :)

SherlockNoMad commented on torch/_export/serde/serialize.py:
```diff
@@ -1364,6 +1428,34 @@ def deserialize_outputs(self, serialized_node: Node, fx_node: torch.fx.Node):
+                    meta_val.append(self.serialized_name_to_meta[name])
+                    # The derived `getitem` nodes should have the same stacktrace as the
+                    # original `fx_node`
+                    individual_output.meta.update(self.deserialize_metadata(serialized_node.metadata))
```
nit... 

`self.deserialize_metadata(serialized_node.metadata)` is returning the same result in line 1445 and 1454? 
This result can be cached in advance.

SherlockNoMad reviewed (approved):
per comments.

zhxchen17 commented on test/export/test_serialize.py:
```diff
@@ -236,7 +236,7 @@ def _check_graph_nodes(gm1, gm2, _check_meta=True):
                         # Or both are fake tensors lists with one element and with the
                         # same shape/dtype
-                        for v1, v2 in zip(val1, val2):
+                        for v1, v2 in zip(pytree.tree_flatten(val1)[0], pytree.tree_flatten(val2)[0]):
```
pytree.tree_flatten() returns (flattend_data, tree_spec)...

zhxchen17 committed:
[export] Upstream support of (tensor, tensor list) in op returns. (#111857)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/111857

Upstreaming from internal to oss.
Diff: D49710320

Test Plan: buck2 build mode/opt sigmoid/inference/test_gpu:package_gen

Reviewed By: SherlockNoMad

Differential Revision: D50577490

fbshipit-source-id: eaa6c7cc95ca16e85534ace4fa9b5caf22eb79d6

zhxchen17 commented:
@pytorchbot merge -i
